File: 217. Contains Duplicate: O(n).py
# 217. Contains Duplicate
# https://leetcode.com/problems/contains-duplicate/description/

# O(n) algorithm for duplicates in array with hashmap of each number count:
class Solution:
    def containsDuplicate(self, nums: List[int]) -> bool:
        map = {}

        for i in range(len(nums)):
            # If number is in dictionary then it appears twice:
            if nums[i] in map:
                return True

            # Increase number count by one (make map if doesn't exist):
            map[nums[i]] = map.get(nums[i], 0) + 1

        # Return false if there were no duplicates found:
        return False

File: 217. Contains Duplicate: O(n).rs
// 217. Contains Duplicate
// https://leetcode.com/problems/contains-duplicate/description/

// O(n) algorithm for duplicates in array with hashmap of each number count:
impl Solution {
    pub fn contains_duplicate(nums: Vec<i32>) -> bool {
        let mut map = std::collections::HashMap::new();

        for &num in &nums {
            // If number is in hashmap then it appears twice:
            if map.contains_key(&num) {
                return true;
            }

            // Increase number count by one (insert hashmap if not present):
            map.entry(num).and_modify(|count| *count += 1).or_insert(1);
        }

        // Return false if there were no duplicates found:
        false
    }
}
